File: merge_sort.js
/*

MERGE SORT

We're going to need a helper function that solves the first major point from above.  

The algorithm for merge sort is actually really simple.

1.  If there is only one element int he list, it is already sorted.  Return that Array

2. otherwise, divide the list recursively into two halves until it can no more be divided.

3. merge the smaller lists into new list in sorted order.


*/

// function merge(array1, array2) {
//     let merged = [];

//     while(array1.length || array2.length) {
//         let ele1 = array1.length ? array1[0] : Infinity;
//         let ele2 = array2.length ? array2[0] : Infinity;

//         let next;
//         if(ele1 < ele2) {
//             next = array1.shift();
//         } else {
//             next = array2.shift();
//         }

//         merged.push(next)
//     }

//     return merged
// }

// function mergeSort(array) {
//     if(array.length <= 1) return array;

//     let midIdx = Math.floor(array.length / 2);
//     let leftHalf = array.slice(0, midIdx);
//     let rightHalf = array.slice(midIdx);

//     let sortedLeft = (mergeSort(leftHalf));
//     let sortedRight = (mergeSort(rightHalf));

//     return merge(sortedLeft, sortedRight)
// }

const merge = (arr1, arr2) => {
    let merged = [];

    while(arr1.length || arr2.length) {
        let next;
        let arr1FirstEle = arr1.length ? arr1[0] : Infinity
        let arr2FirstEle = arr2.length ? arr2[0] : Infinity

        if(arr1FirstEle > arr2FirstEle) next = arr2.shift()
        else next = arr1.shift()

        merged.push(next)
    }

    return merged
}

const mergeSort = (array) => {
    if(array.length <= 1) return array
    
    const midIdx = Math.floor(array.length / 2)

    const leftArr = array.slice(0, midIdx)
    const rightArr = array.slice(midIdx)

    const leftMerge = mergeSort(leftArr)
    const rightMerge = mergeSort(rightArr)

    return merge(leftMerge, rightMerge)
}

console.log(mergeSort([10, -2, 1, 0, -14, 4, 9, 6, 14, -40, 43, 41, -21]))
// console.log(merge([1,3,5,7,9], [2,4,6,8,10]))

File: merge_sort.py
def merge(arr1, arr2):
    merge = []

    arr1Len = len(arr1)
    arr2Len = len(arr2)

    while(arr1Len != 0 or arr2Len != 0):
        arr1First = None
        arr2First = None
        nextUp = None

        if arr1Len != 0:
            arr1First = arr1[0]
        else:
            arr1First = float('inf')
        
        if arr2Len != 0:
            arr2First = arr2[0]
        else:
            arr2First = float('inf')

        if arr1First < arr2First:
            nextUp = arr1.pop(0)
            arr1Len -= 1
        else:
            nextUp = arr2.pop(0)
            arr2Len -= 1

        merge.append(nextUp)

    return merge

def mergeSort(arr):
    length = len(arr)
    if length <= 1:
        return arr

    midIdx = int(length / 2)

    leftArr = arr[0:midIdx]
    rightArr = arr[midIdx:]

    leftSorted = mergeSort(leftArr)
    rightSorted = mergeSort(rightArr)

    return merge(leftSorted, rightSorted)


# print(merge([1, 3, 5, 7, 9], [2, 4, 6, 8, 10]))
print(mergeSort([10, -2, 1, 0, -14, 4, 9, 6, 14, -40, 43, 41, -21]))


